worker/externalcontrollerupdater/externalcontrollerupdater.go

```diff
@@ -0,0 +1,266 @@
+						Site: &cw.catacomb,
+						Work: cw.loop,
+					}); err != nil {
+						return nil, errors.Trace(err)
+					}
+					return &cw, nil
+				})
+			}
+		}
+	}
+}
+
+// controllerWatcher is a worker that watches for changes to the external
+// controller with the given tag. The external controller must be known
+// to the local controller.
+type controllerWatcher struct {
+	catacomb catacomb.Catacomb
+
+	tag                                names.ControllerTag
+	setExternalControllerInfo          func(crossmodel.ControllerInfo) error
+	externalControllerInfo             func(controllerUUID string) (*crossmodel.ControllerInfo, error)
+	newExternalControllerWatcherClient NewExternalControllerWatcherClientFunc
+}
+
+// Kill is part of the worker.Worker interface.
+func (w *controllerWatcher) Kill() {
+	w.catacomb.Kill(nil)
+}
+
+// Wait is part of the worker.Worker interface.
+func (w *controllerWatcher) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *controllerWatcher) loop() error {
+	// We get the API info from the local controller initially.
+	info, err := w.externalControllerInfo(w.tag.Id())
+	if err != nil {
+		return errors.Annotate(err, "getting cached external controller info")
+	}
+	logger.Debugf("controller info for controller %q: %v", w.tag.Id(), info)
+
+	var nw watcher.NotifyWatcher
+	var client ExternalControllerWatcherClientCloser
+	defer func() {
+		if client != nil {
+			client.Close()
+		}
+	}()
+
+	for {
+		if client == nil {
+			apiInfo := &api.Info{
+				Addrs:  info.Addrs,
+				CACert: info.CACert,
+				Tag:    names.NewUserTag(api.AnonymousUsername),
+			}
+			client, err = w.newExternalControllerWatcherClient(apiInfo)
+			if err != nil {
+				return errors.Annotate(err, "getting external controller client")
+			}
+			nw, err = client.WatchControllerInfo()
+			if err != nil {
+				return errors.Annotate(err, "watching external controller")
+			}
+			w.catacomb.Add(nw)
+		}
+
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case _, ok := <-nw.Changes():
+			if !ok {
+				return w.catacomb.ErrDying()
+			}
+
+			newInfo, err := client.ControllerInfo()
+			if err != nil {
+				return errors.Annotate(err, "getting external controller info")
+			}
+			if reflect.DeepEqual(newInfo.Addrs, info.Addrs) {
+				continue
+			}
+			// API addresses have changed. Save the details to the
+			// local controller and stop the existing notify watcher
+			// and set it to nil, so we'll restart it with the new
+			// addresses.
+			info.Addrs = newInfo.Addrs
+			if err := w.setExternalControllerInfo(crossmodel.ControllerInfo{
+				ControllerTag: w.tag,
+				Alias:         info.Alias,
+				Addrs:         info.Addrs,
+				CACert:        info.CACert,
+			}); err != nil {
+				return errors.Annotate(err, "caching external controller info")
+			}
+			logger.Infof("new controller info for controller %q: %v", w.tag.Id(), info)
+			if err := worker.Stop(nw); err != nil {
```

Comment:
Should we also remove nw from the catacomb? Otherwise we are just adding reference to the catacomb each time we create the client again and get a new watcher?

Reply:
the catacomb will remove the worker from its management when the worker stops